rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+
+##### Checkpoint Maintenance
+
+If the reader is opened without a checkpoint, it will periodically re-establish a checkpoint by re-running the initialization process described above, with the following modifications:
+- It will delete the checkpoint that it had previously created (when we support range scans, we’ll need to reference count the checkpoint).
+- It will purge any in-memory tables with content from WAL SSTs lower than the new `wal_id_last_compacted`.
+
+#### Garbage Collector
+
+The GC task will be modified to handle soft deletes and manage checkpoints/clones. The GC tasks's algorithm will now look like the following:
+1. Read the current manifest. If no manifest is found, exit.
+2. If `destroyed_at_s` is set, and the current time is after `destroyed_at_s` + `db_delete_grace`, and there are no active checkpoints, then:
+    1. If the db is a clone, update the source db's manifest by deleting the checkpoint.
+    2. Delete all object's under the db's path.
+    3. Exit.
+3. Garbage collect expired checkpoints
+    1. Find any checkpoints that have expired and remove them
+    2. Write the new manifest version with checkpoints deleted. If CAS fails, go back to 1 to reload checkpoints that may be refreshed or added.
+4. Delete garbage manifests. This includes any manifest that is older than `min_age` and is not referenced by a checkpoint.
+5. Read all manifests referenced by a checkpoint. Call this set M
+6. Clean up WAL SSTs.
+    1. Let C be the lowest id of all `wal_id_last_compacted` in M.
+    2. Delete all WAL SSTs with id lower than C and age older than `min_age`
+7. Clean up SSTs.
+    1. Let S be the set of all SSTs from all manifests in M.
+    2. Delete all SSTs not in M with age older than `min_age`
+8. Detach the clone if possible.
+    1. If the DB instance is a clone, and it's manifest and contained checkpoints no longer references any SSTs from its `source_checkpoint` (we can tell this if all SSTs are under the current db's path), then detach it:
+        1. Update the source db's manifest by removing the checkpoint.
+        2. Update the db's manifest by removing the `source_checkpoint` field.
+    
+Observe that users can now configure a single GC process that can manage GC for multiple databases that use soft deletes. Whenever a new database is created, the user needs to spawn a new GC task for that database. When the GC task completes deleting a database, then the task exits. For now, it's left up to the user to spawn GC tasks for databases that they have created.
```

Comment:
can you expand more? How would we configure GC process to handle multiple paths?

Reply:
I actually wasn't proposing an explicit mechanism for doing this as part of the RFC. I'm just pointing out that as a user you should be able to spawn GC tasks by creating instances of `GarbageCollectorOrchestrator` for each db and calling `run` when a new db instance is created. I suppose we could also come up with some discovery mechanism for detecting dbs under some path and spawning the tasks automatically, but I didn't feel like this was necessary to bring in scope for this RFC.